packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/categories/ExpoNotificationCategoriesModule.java

```diff
@@ -0,0 +1,163 @@
+  private static final String IDENTIFER = "identifier";
+  private static final String BUTTON_TITLE = "buttonTitle";
+  private static final String OPTIONS = "options";
+  private static final String OPENS_APP_TO_FOREGROUND = "opensAppToForeground";
+  private static final String TEXT_INPUT_OPTIONS = "textInput";
+  private static final String SUBMIT_BUTTON_TITLE = "submitButtonTitle";
+  private static final String PLACEHOLDER = "placeholder";
+
+  public ExpoNotificationCategoriesModule(Context context) {
+    super(context);
+  }
+
+  @Override
+  public String getName() {
+    return EXPORTED_NAME;
+  }
+
+  @ExpoMethod
+  public void getNotificationCategoriesAsync(final Promise promise) {
+    BaseNotificationsService.enqueGetCategories(getContext(), new ResultReceiver(null) {
+      @Override
+      protected void onReceiveResult(int resultCode, Bundle resultData) {
+        super.onReceiveResult(resultCode, resultData);
+        Collection<NotificationCategory> categories = resultData.getParcelableArrayList(BaseNotificationsService.CATEGORIES_KEY);
+        if (resultCode == BaseNotificationsService.SUCCESS_CODE && categories != null) {
+          promise.resolve(serializeCategories(categories));
+        } else {
+          Exception e = resultData.getParcelable(BaseNotificationsService.EXCEPTION_KEY);
+          promise.reject("ERR_CATEGORIES_FETCH_FAILED", "A list of notification categories could not be fetched.", e);
+        }
+      }
+    });
+  }
+
+  @ExpoMethod
+  public void setNotificationCategoryAsync(final String identifier, List<HashMap<String, Object>> actionArguments, final Promise promise) {
+    List <NotificationAction> actions = new ArrayList();
+    for (HashMap<String, Object> actionMap : actionArguments) {
+      HashMap<String, Object> actionOptions = getActionOptions(actionMap);
+      HashMap<String, Object> textInputOptions = getTextInputOptions(actionMap);
+      if (textInputOptions != null) {
+        actions.add(new TextInputNotificationAction(getActionIdentifier(actionMap), getActionButtonTitle(actionMap), 
+          getOptionOpensAppToForeground(actionOptions), getSubmitButtonTitle(textInputOptions), getPlaceholder(textInputOptions)));
+      } else {
+        actions.add(new NotificationAction(getActionIdentifier(actionMap), getActionButtonTitle(actionMap), getOptionOpensAppToForeground(actionOptions)));
+      }
+    }
+
+    if (actions.isEmpty()) {
+      throw new InvalidArgumentException("Invalid arguments provided for notification category. Must provide at least one action.");
+    }
+    BaseNotificationsService.enqueSetCategory(getContext(), new NotificationCategory(identifier, actions), new ResultReceiver(null) {
+      @Override
+      protected void onReceiveResult(int resultCode, Bundle resultData) {
+        super.onReceiveResult(resultCode, resultData);
+        NotificationCategory newCategory = resultData.getParcelable(BaseNotificationsService.CATEGORIES_KEY);
```

Comment:
maybe we should reject the promise of the category is null? (an error occurred?) what do you think?

Reply:
yeah, I agree, although [sharedpreferences' commit](https://developer.android.com/reference/android/content/SharedPreferences.Editor.html?hl=en#commit()) won't throw an error, so we can't pass along an error message. But we should throw if this fails anyways